### PR TITLE
540 add python decorators to centralize broiler plate code such as type checking

### PR DESCRIPTION
**Pull request by hokiegeek2**
This PR contains a Python Decorator pattern implementation that centralizes and standardizes type checking for pdarrays. The same decorator can be used in a variety of situations to help ensure we utilize consistent, uniform logic that is used in 2..n places within Arkouda

### CODE REVIEW

File: arkouda/sorting.py [hunk 1]
```diff
@@ -7,13 +9,13 @@
 
 __all__ = ["argsort", "coargsort", "sort"]
 
-def argsort(pda : Union[pdarray,Strings]) -> pdarray:
+def argsort(pda : Union[pdarray,Strings,'Categorical']) -> pdarray:
```
Comment (mhmerrill) [inline arkouda/sorting.py:12]: @hokiegeek2 did you need a `Categorical = ForwardRef('Categorical')` in this file

Comment (hokiegeek2) [reply arkouda/sorting.py:12]: @mhmerrill No, I tried to do that but circular dependencies are preventing putting in the actual Categorical classname, so no, not needed in this file.

Comment (mhmerrill) [reply arkouda/sorting.py:12]: ok

**Review by mhmerrill — APPROVED** [review]
things look ok to me

**Comment by mhmerrill** [comment]
I just realized that the `README.md` needs to be updated with the new requirement of Python 3.7 or greater

**Comment by hokiegeek2** [comment]
Argh, yeah, that's correct.

**Comment by mhmerrill** [comment]
> Argh, yeah, that's correct.

I am going to open another PR to update the README.md and fix a few things

**Pull request merged**